pkg/controller/machine-set-boot-image/platform_helpers.go

```diff
@@ -192,9 +209,348 @@ func reconcileLibvirt(machineSet *machinev1beta1.MachineSet, _ *corev1.ConfigMap
+	u, err := client.Client.ParseURL(rawURL)
+	if err != nil {
+		return err
+	}
+	url := u.String()
+	c := client.Client.Client
+
+	param := soap.Upload{
+		ContentLength: size,
+	}
+
+	if item.Create {
+		param.Method = "PUT"
+		param.Headers = map[string]string{
+			"Overwrite": "t",
+		}
+	} else {
+		param.Method = "POST"
+		param.Type = "application/x-vnd.vmware-streamVmdk"
+	}
+
+	pr := &ProgressReader{f, func(r int64) {
+		incrementTotalBytesRead(totalBytesRead, r)
+	}}
+	f = pr
+
+	req, err := http.NewRequest(param.Method, url, f)
+	if err != nil {
+		return err
+	}
+
+	req = req.WithContext(ctx)
+	req.ContentLength = param.ContentLength
+	req.Header.Set("Content-Type", param.Type)
+
+	for k, v := range param.Headers {
+		req.Header.Add(k, v)
+	}
+	if param.Ticket != nil {
+		req.AddCookie(param.Ticket)
+	}
+
+	res, err := c.Client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer func(Body io.ReadCloser) {
+		_ = Body.Close()
+	}(res.Body)
+
+	switch res.StatusCode {
+	case http.StatusOK:
+	case http.StatusCreated:
+	default:
+		err = errors.New(res.Status)
+	}
+	return err
+}
+
+func findAndUploadDiskFromOva(client *govmomi.Client, ovaFile io.Reader, diskName string, ovfFileItem types.OvfFileItem, deviceObj types.HttpNfcLeaseDeviceUrl, currBytesRead *int64) error {
+	ovaReader := tar.NewReader(ovaFile)
+	for {
+		fileHdr, err := ovaReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if fileHdr.Name == diskName {
+			err = upload(context.Background(), client, ovfFileItem, ovaReader, deviceObj.Url, ovfFileItem.Size, currBytesRead)
+			if err != nil {
+				return fmt.Errorf("error while uploading the file %s %s", diskName, err)
+			}
+			return nil
+		}
+	}
+	return fmt.Errorf("disk %s not found inside ova", diskName)
+}
+
+func uploadOvaDisksFromLocal(client *govmomi.Client, filePath string, ovfFileItem types.OvfFileItem, deviceObj types.HttpNfcLeaseDeviceUrl, currBytesRead *int64) error {
+	diskName := ovfFileItem.Path
+	ovaFile, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+	defer func(ovaFile *os.File) {
+		_ = ovaFile.Close()
+	}(ovaFile)
+
+	err = findAndUploadDiskFromOva(client, ovaFile, diskName, ovfFileItem, deviceObj, currBytesRead)
+	return err
+}
+
+func getOvfDescriptorFromOva(ovaFile io.Reader) (string, error) {
+	ovaReader := tar.NewReader(ovaFile)
+	for {
+		fileHdr, err := ovaReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return "", err
+		}
+		if strings.HasSuffix(fileHdr.Name, ".ovf") {
+			content, _ := io.ReadAll(ovaReader)
+			ovfDescriptor := string(content)
+			return ovfDescriptor, nil
+		}
+	}
+	return "", fmt.Errorf("ovf file not found inside the ova")
+}
+
+func createNewVMTemplateWithName(name string, streamData *stream.Stream, providerSpec *machinev1beta1.VSphereMachineProviderSpec, arch string) error {
```

Comment:
I think _maybe_ the infrastructure spec should be used first and the machine spec should be secondary

https://github.com/openshift/api/blob/c1fdeb0788c16659238d93ec45ce2e798c14eb88/config/v1/types_infrastructure.go#L1243-L1315